src/services/huaweicloud/obs.rs

```diff
@@ -0,0 +1,324 @@
+    writeln!(&mut s, "{}", get_or_default(&h, &CONTENT_MD5.parse()?)?)?;
+    writeln!(&mut s, "{}", get_or_default(&h, &CONTENT_TYPE)?)?;
+    writeln!(&mut s, "{}", get_or_default(&h, &DATE)?)?;
+    let canonicalize_header = canonicalize_header(req)?;
+    if !canonicalize_header.is_empty() {
+        writeln!(&mut s, "{}", canonicalize_header)?;
+    }
+    write!(&mut s, "{}", canonicalize_resource(req, cred)?)?;
+
+    debug!("string to sign: {}", &s);
+
+    Ok(s)
+}
+
+/// ## Reference
+///
+/// - [Authentication of Signature in a Header](https://support.huaweicloud.com/intl/en-us/api-obs/obs_04_0010.html)
+fn canonicalize_header(req: &impl SignableRequest) -> Result<String> {
+    let mut headers = req
+        .headers()
+        .iter()
+        // Filter all header that starts with "x-obs-"
+        .filter(|(k, _)| k.as_str().starts_with("x-obs-"))
+        // Convert all header name to lowercase
+        .map(|(k, v)| {
+            (
+                k.as_str().to_lowercase(),
+                v.to_str().expect("must be valid header").to_string(),
+            )
+        })
+        .collect::<Vec<(String, String)>>();
+
+    // Sort via header name.
+    headers.sort_by(|x, y| x.0.cmp(&y.0));
+
+    Ok(headers
+        .iter()
+        // Format into "name:value"
+        .map(|(k, v)| format!("{}:{}", k, v))
+        .collect::<Vec<String>>()
+        // Join via "\n"
+        .join("\n"))
+}
+
+/// ## Reference
+///
+/// - [Authentication of Signature in a Header](https://support.huaweicloud.com/intl/en-us/api-obs/obs_04_0010.html)
+fn canonicalize_resource(req: &impl SignableRequest, _cred: &Credential) -> Result<String> {
+    let bucket_name = parse_bucket_name(req.host());
+
+    let mut params: Vec<(Cow<'_, str>, Cow<'_, str>)> =
+        form_urlencoded::parse(req.query().unwrap_or_default().as_bytes()).collect();
+    // Sort by param name
+    params.sort();
+
+    let mut s = String::new();
+
+    write!(&mut s, "/{}", bucket_name.to_string())?;
+    write!(&mut s, "{}", req.path())?;
+
+    let params_str = params
+        .iter()
+        .map(|(k, v)| format!("{}={}", k, v))
+        .collect::<Vec<String>>()
+        .join("&");
+
+    if !params_str.is_empty() {
+        write!(s, "?{}", params_str)?;
+    }
+
+    Ok(s.to_string())
+}
+
+enum ParseBucketNameResult {
+    CustomBucketDomain(String),
+    DefaultDomainWithBucket(String),
+    DefaultDomainWithoutBucket,
+}
+
+impl ToString for ParseBucketNameResult {
+    fn to_string(&self) -> String {
+        match self {
+            ParseBucketNameResult::CustomBucketDomain(d) => d.clone(),
+            ParseBucketNameResult::DefaultDomainWithBucket(d) => d.clone(),
+            ParseBucketNameResult::DefaultDomainWithoutBucket => "".to_string(),
+        }
+    }
+}
+
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"(([\w\W]+).)?obs.[\w\W]+.myhuaweicloud.com").unwrap();
+}
+
+fn parse_bucket_name(host: &str) -> ParseBucketNameResult {
```

Comment:
I prefer to let the user decide this.

How about exposing an `fn set_bucket(bucket: &str)` to the user?

Reply:
Exposing a set_bucket() is more flexible, but it may be misused.

e.g.: set the domain to `bucket1.obs.cn-north-1.myhuaweicloud.com`, then call `set_bucket('bucket2')`.

To avoid this, we have to check the bucket names in Signer.

Reply:
The user must make sure the input bucket name is correct. We can make this clear in the docs of this function.

Reply:
Should I store the bucket name in Signer when building it?

```rust
    Signer::builder()
        .access_key("ak")
        .secret_key("sk")
        .bucket("databend")
        .build()?;
```

Reply:
fixed